src/objective/multiclass_obj.cu

```diff
@@ -0,0 +1,179 @@
+/*!
+ * Copyright 2015-2018 by Contributors
+ * \file multi_class.cc
+ * \brief Definition of multi-class classification objectives.
+ * \author Tianqi Chen
+ */
+#include <dmlc/omp.h>
+#include <dmlc/parameter.h>
+#include <xgboost/data.h>
+#include <xgboost/logging.h>
+#include <xgboost/objective.h>
+#include <vector>
+#include <algorithm>
+#include <limits>
+#include <utility>
+#include "../common/math.h"
+#include "../common/transform.h"
+
+namespace xgboost {
+namespace obj {
+
+#if defined(XGBOOST_USE_CUDA)
+DMLC_REGISTRY_FILE_TAG(multiclass_obj_gpu);
+#endif
+
+struct SoftmaxMultiClassParam : public dmlc::Parameter<SoftmaxMultiClassParam> {
+  int num_class;
+  int n_gpus;
+  int gpu_id;
+  // declare parameters
+  DMLC_DECLARE_PARAMETER(SoftmaxMultiClassParam) {
+    DMLC_DECLARE_FIELD(num_class).set_lower_bound(1)
+        .describe("Number of output class in the multi-class classification.");
+    DMLC_DECLARE_FIELD(n_gpus).set_default(-1).set_lower_bound(-1)
+        .describe("Number of GPUs to use for multi-gpu algorithms.");
+    DMLC_DECLARE_FIELD(gpu_id)
+        .set_lower_bound(0)
+        .set_default(0)
+        .describe("gpu to use for objective function evaluation");
+  }
+};
+
+class SoftmaxMultiClassObj : public ObjFunction {
+ public:
+  explicit SoftmaxMultiClassObj(bool output_prob)
+  : output_prob_(output_prob) {
+  }
+  void Configure(const std::vector<std::pair<std::string, std::string> >& args) override {
+    param_.InitAllowUnknown(args);
+    CHECK(param_.n_gpus != 0) << "Must have at least one device";  // Default is -1
+    devices_ = GPUSet::All(param_.n_gpus).Normalised(param_.gpu_id);
+    label_correct_.Resize(devices_.IsEmpty() ? 1 : devices_.Size());
+  }
+  void GetGradient(const HostDeviceVector<bst_float>& preds,
+                   const MetaInfo& info,
+                   int iter,
+                   HostDeviceVector<GradientPair>* out_gpair) override {
+    CHECK_NE(info.labels_.Size(), 0U) << "label set cannot be empty";
+    CHECK(preds.Size() == (static_cast<size_t>(param_.num_class) * info.labels_.Size()))
+        << "SoftmaxMultiClassObj: label size and pred size does not match";
+    out_gpair->Resize(preds.Size());
+    label_correct_.Fill(1);
+    const int nclass = param_.num_class;
+    const auto ndata = static_cast<int64_t>(preds.Size() / nclass);
+    common::ReshardAll(out_gpair, GPUDistribution::Block(devices_),
+                       &info.labels_, GPUDistribution::Block(devices_),
+                       &preds, GPUDistribution::Granular(devices_, nclass),
+                       &info.weights_, GPUDistribution::Block(devices_),
```

Comment:
Please avoid this, and just call `Reshard()` on each `HostDeviceVector` individually.